x/gamm/pool-models/balancer/amm.go

```diff
@@ -349,27 +349,44 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	// more complex / don't just alter the state.
 	// We should simplify this logic further in the future, using balancer multi-join equations.
 
+	if tokensIn.Len() == 1 {
```

Comment:
Thanks for this PR @hieuvubk 

I think we might want to delegate this decision to the caller of this function. More specifically, have 2 separate exported methods:
- `CalcMultiTokenInJoinPoolShares`
- `CalcSingleTokenInJoinPoolShares`

However, I'm not sure if defining these 2 methods on the pool interface would make sense. Would these methods also work for stable swap?

Let's wait on @ValarDragon 's feedback on this before making any changes, please.

Reply:
I think this idea makes sense 👍

Reply:
Thanks for the suggestion!
I have a question. In amm pool model, **JoinPool** function calls to **CalcJoinPoolShares** internally, should we separate **JoinPool** to single and multi join or based on tokensIn length to call

Reply:
Good point. Let's keep `JoinPool` as is for now and make it decide which "calc" method to call

For more context and as per https://github.com/osmosis-labs/osmosis/pull/1750, we are considering splitting the `JoinPool` abstraction to `JoinPool` and `JoinPoolNoSwap`.

If we make single and multi-asset join methods and also adopt #1750, there will be 4 different join pool variations.

We can investigate later to determine the best path forward. For now, to limit the scope, let's only deal with the calc methods, please